red_queen/games/applications/run_hamiltonian_sim.py

```diff
@@ -0,0 +1,209 @@
+    -0.2517444632292354,
+    0.867164336543846,
+    0.296282404856135,
+    0.3435823547580954,
+    0.8631471356486622,
+    0.3399761708363991,
+    -0.5357509772547095,
+    -0.4883312979088732,
+    -0.5500163526392827,
+    -0.8385986892423059,
+    -0.6319995154933682,
+    0.44989873651812795,
+]
+K = 3
+W = 10
+T = 0.01
+
+
+# xx gate
+def xx_gate(angle):
+    qc = QuantumCircuit(2, name="xx")
+    for i in range(2):
+        qc.h(i)
+    qc.cx(0, 1)
+    qc.rz(np.pi * angle, 1)
+    qc.cx(0, 1)
+    for i in range(2):
+        qc.h(i)
+
+    return qc
+
+
+# yy gate
+def yy_gate(angle):
+    qc = QuantumCircuit(2, name="yy")
+    for i in range(2):
+        qc.s(i)
+    for i in range(2):
+        qc.h(i)
+    qc.cx(0, 1)
+    qc.rz(np.pi * angle, 1)
+    qc.cx(0, 1)
+    for i in range(2):
+        qc.h(i)
+    for i in range(2):
+        qc.sdg(i)
+
+    return qc
+
+
+# zz_gate
+def zz_gate(angle):
+    qc = QuantumCircuit(2, name="zz")
+    qc.cx(0, 1)
+    qc.rz(np.pi * angle, 1)
+    qc.cx(0, 1)
+
+    return qc
+
+
+# Hamiltonian simulation circuit generator, where:
+#    n = number of qubits
+#    k_order = trotterization order (number of trotter steps)
+#    time = time the simulation runs.
+#    w = strength of disorder
+#    h_x and h_z = disordered fields.
+def hamiltonian_sim(n: int, k_order: int, time, w: int, h_x, h_z):
+    qr = QuantumRegister(n)
+    cr = ClassicalRegister(n)
+    qc = QuantumCircuit(qr, cr, name="Hamiltonian")
+    angle = time / k_order
+
+    # Create binary state of "10101010..."
+    for i in range(0, n, 2):
+        qc.x(qr[i])
+    qc.barrier()
+
+    # Apply k amount of trotterization steps
+    for i in range(k_order):
+
+        # Pauli spin
+        for j in range(n):
+            qc.rx(2 * angle * w * h_x[j], qr[j])
+        for j in range(n):
+            qc.rz(2 * angle * w * h_z[j], qr[j])
+        qc.barrier()
+
+        # Apply xx_gate:
+        for j in range(2):
+            for k in range(j % 2, n - 1, 2):
+                qc.append(xx_gate(angle).to_instruction(), [qr[k], qr[(k + 1) % n]])
+
+        # Apply yy_gate:
+        for j in range(2):
+            for k in range(j % 2, n - 1, 2):
+                qc.append(yy_gate(angle).to_instruction(), [qr[k], qr[(k + 1) % n]])
+
+        # Apply zz_gate:
+        for j in range(2):
+            for k in range(j % 2, n - 1, 2):
+                qc.append(zz_gate(angle).to_instruction(), [qr[k], qr[(k + 1) % n]])
+
+        qc.barrier()
+
+    # Measure all qubits
+    for qubit in range(n):
+        qc.measure(qr[qubit], cr[qubit])
+
+    return qc
+
+
+@pytest.mark.qiskit
+@pytest.mark.parametrize("optimization_level", [0, 1, 2, 3])
+@pytest.mark.parametrize("backend", backends)
+def bench_qiskit_ft(benchmark, optimization_level, backend):
+    shots = 65536
+    # Generate binary string based on number of qubits
+    binary_string = ""
+    for i in range(NUM_QUBITS):
+        binary_string += "1" if i % 2 == 0 else "0"
```

Comment:
This isn't super critical for this application, but as a rule of thumb doing string += like this can be slow for large strings because python is doing copies under the covers. It's normally more efficient to use `io.StringIO` to write to a single buffer and just get the value from the buffer to do iterative string generation like this. However in this case you can just use a generator or a list comprehension like:

```suggestion
    "".join(["1" if i % 2 == 0 else "0" for i in range(NUM_QUBITS)])
```